go/vt/vtgate/engine/join.go

```diff
@@ -216,14 +219,17 @@ func (jn *Join) Inputs() ([]Primitive, []map[string]any) {
 	return []Primitive{jn.Left, jn.Right}, nil
 }
 
-func joinFields(lfields, rfields []*querypb.Field, cols []int) []*querypb.Field {
-	fields := make([]*querypb.Field, len(cols))
-	for i, index := range cols {
+func (jn *Join) joinFields(lfields, rfields []*querypb.Field) []*querypb.Field {
+	fields := make([]*querypb.Field, len(jn.Cols))
+	for i, index := range jn.Cols {
 		if index < 0 {
-			fields[i] = lfields[-index-1]
-			continue
+			fields[i] = lfields[-index-1].CloneVT()
+		} else {
+			fields[i] = rfields[index-1].CloneVT()
+		}
+		if len(jn.ColNames) > i && jn.ColNames[i] != "" {
+			fields[i].Name = jn.ColNames[i]
 		}
```

Comment:
nit: this is an all-or-nothing case i.e. ColNames either exists for some/all or nothing.
For nothing case should we not use the old method?

Also when ColNames are present, we can check that first and continue on the next iteration.